File: lectures/025bintrees/bintree.py
#   Bintree from EOPL p. 46 in Python
#   Illustrating object oriented programming.
#   Geoffrey Matthews
#   Tested using python on linux
#   Thu Apr  4 12:16:36 PST 2002     


class BinaryTree:
  def leafSum():
    return 0
  def printTree(self):
    self.printTreeIndented(0)
  def printTreeIndented(self,depth):
    print("Nonexistent tree.")
    
class Leaf(BinaryTree):
  def __init__(self,n):
    self.datum = n
  def leafSum(self):
    return self.datum
  def printTreeIndented(self,depth):
    for i in range(depth):
      print(" ", end='')
    print(self.datum)
  
class InteriorNode(BinaryTree):
  def __init__(self,k,lf,rt):
    self.key = k;
    self.left = lf;
    self.right = rt;
  def leafSum(self):
    return self.left.leafSum() + self.right.leafSum()
  def printTreeIndented(self,depth):
    for i in range(depth):
      print(" ", end='')
    print(self.key)
    self.left.printTreeIndented(depth+1)
    self.right.printTreeIndented(depth+1)

x = InteriorNode("Fuzzy",
                 InteriorNode("Wuzzy",
                              Leaf(11),
                              InteriorNode("Was",
                                           Leaf(22),
                                           Leaf(33))),
                 InteriorNode("A",
                              InteriorNode("Bear",
                                           Leaf(44),
                                           Leaf(55)),
                              Leaf(66)))

x.printTree()
print("The sum is:", x.leafSum())

File: lectures/025bintrees/bintree.java
/* Bintree from EOPL p. 46 in Java
   Illustrating object oriented programming.
   Geoffrey Matthews
   Tested using javac on linux
   Thu Apr  4 10:54:55 PST 2002  
*/

class BinaryTree {
    public int leafSum() {
	return 0;
    }
    public void printTree() {
	printTreeIndented(0);
    }
    public void printTreeIndented(int depth) {
	System.out.println("Nonexistent tree.\n");
    }
}

class Leaf extends BinaryTree {
    public Leaf(int n) {
	datum = n;
    }
    public void printTreeIndented(int depth) {
	for (int i = 0; i < depth; i++)
	    System.out.print("  ");
	System.out.println(datum);
    }
    public int leafSum () {
	return datum;
    }
    private int datum;
}

class InteriorNode extends BinaryTree {
    public InteriorNode(String k, BinaryTree lf, BinaryTree rt) {
	key = k;
	left = lf;
	right = rt;
    }
    public void printTreeIndented(int depth) {
	for (int i = 0; i < depth; i++)
	    System.out.print("  ");
	System.out.println(key);
	left.printTreeIndented(depth+1);
	right.printTreeIndented(depth+1);
    }	
    public int leafSum () {
	return left.leafSum() + right.leafSum();
    }
    private String key;
    private BinaryTree left;
    private BinaryTree right;
}
	
class bintree {
    public static void main(String[] args) {
	BinaryTree x =
	    new InteriorNode(
		"Fuzzy",
		new InteriorNode(
		    "Wuzzy",
		    new Leaf(11),
		    new InteriorNode(
			"Was",
			new Leaf(22),
			new Leaf(33))),
		new InteriorNode(
		    "A",
		    new InteriorNode(
			"Bear",
			new Leaf(44),
			new Leaf(55)),
		    new Leaf(66)));
	x.printTree();
	System.out.println("The sum is: " + x.leafSum());
    }
}
	
